File: 2016/day01_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import Literal

type Direction = Literal["N", "S", "W", "E"]
type Position = tuple[int, int, Direction]
type Turn = Literal["L", "R"]
type Instruction = tuple[Turn, int]


def solution(content: str) -> None:
    instructions: list[Instruction] = [
        parse(instruction) for instruction in content.split(", ")
    ]
    current: Position = (0, 0, "N")
    for instruction in instructions:
        current = move(current, instruction)

    print(abs(current[0]) + abs(current[1]))


def turn(position: Position, instruction: Instruction) -> Position:
    if position[2] == "E":
        return (position[0], position[1], "N" if instruction[0] == "L" else "S")
    if position[2] == "W":
        return (position[0], position[1], "S" if instruction[0] == "L" else "N")
    if position[2] == "N":
        return (position[0], position[1], "W" if instruction[0] == "L" else "E")
    if position[2] == "S":
        return (position[0], position[1], "E" if instruction[0] == "L" else "W")
    raise ValueError()


def walk(position: Position, instruction: Instruction) -> Position:
    if position[2] == "N":
        return (position[0], position[1] + instruction[1], position[2])
    if position[2] == "S":
        return (position[0], position[1] - instruction[1], position[2])
    if position[2] == "E":
        return (position[0] + instruction[1], position[1], position[2])
    if position[2] == "W":
        return (position[0] - instruction[1], position[1], position[2])
    raise ValueError()


def move(position: Position, instruction: Instruction) -> Position:
    return walk(turn(position, instruction), instruction)


def parse(element: str) -> Instruction:
    return ("L" if element[0] == "L" else "R", int(element[1:]))


def main() -> None:
    with open("day01.dat", "r", encoding="utf8") as file:
        solution(file.read())


if __name__ == "__main__":
    main()

File: 2016/day01-1.ts
type Direction = "N" | "S" | "W" | "E";
type Turn = "L" | "R";
type Instruction = {
  turn: Turn;
  blocks: number;
};
type Position = {
  x: number;
  y: number;
  direction: Direction;
};

function solution(content: string) {
  const final = content.split(", ").map(parse).reduce(move, {
    x: 0,
    y: 0,
    direction: "N",
  });
  console.log(Math.abs(final.x) + Math.abs(final.y));
}

function turn(position: Position, instruction: Instruction): Position {
  if (position.direction === "N")
    return { ...position, direction: instruction.turn === "L" ? "W" : "E" };
  if (position.direction === "S")
    return { ...position, direction: instruction.turn === "L" ? "E" : "W" };
  if (position.direction === "E")
    return { ...position, direction: instruction.turn === "L" ? "N" : "S" };
  if (position.direction === "W")
    return { ...position, direction: instruction.turn === "L" ? "S" : "N" };
  return position;
}

function walk(position: Position, instruction: Instruction): Position {
  if (position.direction === "N")
    return { ...position, y: position.y + instruction.blocks };
  if (position.direction === "S")
    return { ...position, y: position.y - instruction.blocks };
  if (position.direction === "E")
    return { ...position, x: position.x + instruction.blocks };
  if (position.direction === "W")
    return { ...position, x: position.x - instruction.blocks };
  return position;
}

function move(position: Position, instruction: Instruction): Position {
  return walk(turn(position, instruction), instruction);
}

function parse(item: string): Instruction {
  return {
    turn: item[0] === "L" ? "L" : "R",
    blocks: parseInt(item.substring(1)),
  };
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution);
